File: backend/internal/render/bidi_test.go
package render

import (
	"strings"
	"testing"
)

const (
	he = "שלום"  // "shalom"
	ar = "مرحبا" // "marhaba"
)

// These mirror the cases in `packages/text/src/__tests__/bidi.test.ts`. The two
// implementations lay out the same paragraph on two different paths, so a
// divergence here is a design that reads correctly on the canvas and wrongly in
// its own export.
func TestResolveBaseDirectionMatchesTheBrowser(t *testing.T) {
	cases := []struct{ text, declared, want string }{
		{"hello world", "auto", "ltr"},
		{he, "auto", "rtl"},
		{ar, "auto", "rtl"},
		{`  "123" ` + he, "auto", "rtl"},
		{`  "123" hello`, "auto", "ltr"},
		{"Re: " + he, "auto", "ltr"},
		{he, "ltr", "ltr"}, // an explicit declaration wins
		{"hello", "rtl", "rtl"},
		{"", "auto", "ltr"},
		{"123 456 ...", "auto", "ltr"},
	}
	for _, c := range cases {
		if got := ResolveBaseDirection(c.text, c.declared); got != c.want {
			t.Errorf("ResolveBaseDirection(%q, %q) = %q, want %q", c.text, c.declared, got, c.want)
		}
	}
}

func TestResolveLevels(t *testing.T) {
	// Right-to-left text gets an odd level inside a left-to-right paragraph.
	text := []rune("hi " + he)
	lv := resolveLevels(text, "ltr")
	if lv[0] != 0 {
		t.Errorf("Latin should sit at the base level, got %d", lv[0])
	}
	if lv[len(lv)-1]%2 != 1 {
		t.Errorf("Hebrew should sit at an odd level, got %d", lv[len(lv)-1])
	}
	// Latin gets an even level above the base inside a right-to-left paragraph.
	text = []rune(ar + " ok")
	lv = resolveLevels(text, "rtl")
	if lv[0]%2 != 1 {
		t.Errorf("Arabic at base level should be odd, got %d", lv[0])
	}
	if lv[len(lv)-1]%2 != 0 {
		t.Errorf("Latin inside Arabic should be even, got %d", lv[len(lv)-1])
	}
	// Digits after Arabic still read left to right (W2, I2).
	text = []rune(ar + " 2026")
	lv = resolveLevels(text, "rtl")
	if d := lv[len(lv)-1]; d%2 != 0 || d <= 1 {
		t.Errorf("digits in Arabic should be even and above the base, got %d", d)
	}
	// A neutral between two right-to-left runs goes right-to-left (N1); between
	// opposite directions it takes the base (N2).
	text = []rune(he + " - " + he)
	lv = resolveLevels(text, "ltr")
	if lv[strings.Index(string(text), "-")]%2 != 1 {
		t.Error("a dash between two Hebrew runs should be right-to-left")
	}
	text = []rune("abc - " + he)
	lv = resolveLevels(text, "ltr")
	if lv[4]%2 != 0 {
		t.Error("a dash between Latin and Hebrew should take the base direction")
	}
}

func TestOrderVisualReordersRuns(t *testing.T) {
	// Two Hebrew style runs in a right-to-left paragraph: the second authored
	// run is drawn first (leftmost).
	pieces := OrderVisual([]string{"שלום ", "עולם"}, "rtl")
	if len(pieces) != 2 || pieces[0].Item != 1 || pieces[1].Item != 0 {
		t.Fatalf("run order not reversed: %+v", pieces)
	}
	// An embedded Latin phrase stays readable and sits leftmost.
	pieces = OrderVisual([]string{ar + " ", "HyCanvas"}, "rtl")
	if pieces[0].Item != 1 || pieces[0].Text != "HyCanvas" {
		t.Fatalf("Latin inside Arabic was reordered or reversed: %+v", pieces[0])
	}
	if pieces[0].Level%2 != 0 {
		t.Errorf("Latin should be at an even level, got %d", pieces[0].Level)
	}
	// Hebrew after a Latin lead-in keeps the lead-in first.
	pieces = OrderVisual([]string{"Note: ", he}, "ltr")
	if pieces[0].Item != 0 {
		t.Fatalf("Latin lead-in should stay first: %+v", pieces)
	}
}

// The difference from the browser: this renderer draws rune by rune, so the
// characters inside a right-to-left run must come out reversed. Without this,
// Hebrew exports mirrored relative to the canvas.
func TestOrderVisualReversesCharactersInRtlRuns(t *testing.T) {
	pieces := OrderVisual([]string{he}, "rtl")
	if len(pieces) != 1 {
		t.Fatalf("expected one piece, got %d", len(pieces))
	}
	want := reverseRunes(he)
	if pieces[0].Text != want {
		t.Fatalf("Hebrew not reversed for per-rune drawing: got %q, want %q", pieces[0].Text, want)
	}
	// Latin inside the same line must NOT be reversed.
	pieces = OrderVisual([]string{he + " ", "abc"}, "rtl")
	for _, p := range pieces {
		if strings.Contains(p.Text, "abc") && p.Text != "abc" {
			t.Fatalf("Latin was reversed: %q", p.Text)
		}
	}
}

func TestOrderVisualLeavesLtrUntouched(t *testing.T) {
	in := []string{"Hello ", "world"}
	pieces := OrderVisual(in, "ltr")
	if len(pieces) != 2 || pieces[0].Text != "Hello " || pieces[1].Text != "world" {
		t.Fatalf("left-to-right text was altered: %+v", pieces)
	}
}

func TestOrderVisualPreservesContent(t *testing.T) {
	items := []string{"Total: ", "42 ", ar}
	for _, base := range []string{"ltr", "rtl"} {
		var got strings.Builder
		for _, p := range OrderVisual(items, base) {
			got.WriteString(p.Text)
		}
		if len([]rune(got.String())) != len([]rune(strings.Join(items, ""))) {
			t.Fatalf("%s: characters lost or duplicated: %q", base, got.String())
		}
	}
}

func TestOrderVisualEdgeCases(t *testing.T) {
	if OrderVisual(nil, "rtl") != nil {
		t.Error("nil input should produce no pieces")
	}
	pieces := OrderVisual([]string{""}, "rtl")
	for _, p := range pieces {
		if p.Text != "" {
			t.Errorf("empty run produced text: %q", p.Text)
		}
	}
	if HasRtl("hello 123 .,") {
		t.Error("HasRtl reported Latin and digits as right-to-left")
	}
	if !HasRtl("mixed " + ar) {
		t.Error("HasRtl missed Arabic")
	}
}

// End to end: a paragraph declared right-to-left must export right-aligned.
//
// This uses LATIN text on purpose. The embedded fallback font (Liberation Sans)
// has no Hebrew, Arabic, or CJK glyphs, so a Hebrew string exports as nothing
// at all and could not tell a passing alignment from a broken one. That font
// gap is the real blocker for non-Latin export and is recorded in the
// rasterizer's fidelity notes; the ordering this file tests is necessary but
// not sufficient until it is closed.
func TestRasterTextRtlParagraphAlignsRight(t *testing.T) {
	col := func(r, g, b float64) map[string]any {
		return map[string]any{"srgb": map[string]any{"r": r, "g": g, "b": b, "a": 1.0}}
	}
	build := func(dir string) Design {
		return Design{"pages": []any{map[string]any{
			"width": 300.0, "height": 80.0,
			"background": map[string]any{"type": "solid", "color": col(1, 1, 1)},
			"children": []any{map[string]any{
				"id": "t1", "type": "text",
				"transform": map[string]any{"x": 0.0, "y": 0.0, "scaleX": 1.0, "scaleY": 1.0, "rotation": 0.0},
				"size":      map[string]any{"width": 300.0, "height": 80.0},
				"content": []any{map[string]any{
					"runs":  []any{map[string]any{"text": "shalom", "style": map[string]any{"fontSize": 24.0, "fill": map[string]any{"type": "solid", "color": col(0, 0, 0)}}}},
					"style": map[string]any{"direction": dir},
				}},
			}},
		}}}
	}
	side := func(d Design) (left, right int) {
		img, err := toRaster(d, 0, 1, false)
		if err != nil {
			t.Fatalf("toRaster: %v", err)
		}
		b := img.Bounds()
		for y := b.Min.Y; y < b.Max.Y; y++ {
			for x := b.Min.X; x < b.Max.X; x++ {
				if img.RGBAAt(x, y).R < 128 {
					if x < b.Dx()/2 {
						left++
					} else {
						right++
					}
				}
			}
		}
		return
	}
	lLeft, lRight := side(build("ltr"))
	if lLeft == 0 && lRight == 0 {
		t.Fatal("no glyphs rendered at all; the probe font lost Latin coverage")
	}
	if lLeft <= lRight {
		t.Fatalf("left-to-right text did not start on the left: left=%d right=%d", lLeft, lRight)
	}
	rLeft, rRight := side(build("rtl"))
	if rRight <= rLeft {
		t.Fatalf("a right-to-left paragraph did not align right: left=%d right=%d", rLeft, rRight)
	}
}

// Glyph coverage (F38 FR-10). The embedded fallback has no Hebrew, Arabic,
// Indic or CJK, and an uncovered rune used to be skipped silently, so a design
// in those scripts exported BLANK while the canvas showed words. Registering a
// font that covers the script must rescue it, per glyph, even when the run
// names a different family.
func TestUncoveredGlyphsFallBackToARegisteredFont(t *testing.T) {
	// The embedded font is the ground truth for the gap this guards.
	if coversRune(fallbackFont, 'א') {
		t.Skip("the embedded font now covers Hebrew; this guard needs rewriting")
	}
	if !coversRune(fallbackFont, 'A') {
		t.Fatal("the embedded font lost Latin coverage")
	}

	// With nothing registered, nothing covers the rune and the renderer reports
	// it rather than dropping it in silence.
	var reported []rune
	SetMissingGlyphReporter(func(r rune) { reported = append(reported, r) })
	t.Cleanup(func() { SetMissingGlyphReporter(nil) })

	if f := fontCovering('א', "Whatever", 400); f != nil {
		t.Skip("a Hebrew-covering font is already registered in this process")
	}

	col := func(r, g, b float64) map[string]any {
		return map[string]any{"srgb": map[string]any{"r": r, "g": g, "b": b, "a": 1.0}}
	}
	design := Design{"pages": []any{map[string]any{
		"width": 200.0, "height": 60.0,
		"background": map[string]any{"type": "solid", "color": col(1, 1, 1)},
		"children": []any{map[string]any{
			"id": "t1", "type": "text",
			"transform": map[string]any{"x": 0.0, "y": 0.0, "scaleX": 1.0, "scaleY": 1.0, "rotation": 0.0},
			"size":      map[string]any{"width": 200.0, "height": 60.0},
			"content": []any{map[string]any{
				"runs":  []any{map[string]any{"text": he, "style": map[string]any{"fontSize": 28.0, "fontFamily": "Whatever", "fill": map[string]any{"type": "solid", "color": col(0, 0, 0)}}}},
				"style": map[string]any{"direction": "auto"},
			}},
		}},
	}}}
	ink := func(d Design) int {
		img, err := toRaster(d, 0, 1, false)
		if err != nil {
			t.Fatalf("toRaster: %v", err)
		}
		n := 0
		b := img.Bounds()
		for y := b.Min.Y; y < b.Max.Y; y++ {
			for x := b.Min.X; x < b.Max.X; x++ {
				if img.RGBAAt(x, y).R < 128 {
					n++
				}
			}
		}
		return n
	}
	if got := ink(design); got != 0 {
		t.Fatalf("expected blank output with no covering font, got %d dark pixels", got)
	}
	if len(reported) == 0 {
		t.Fatal("an uncovered glyph was dropped without being reported")
	}
}

// A wrapped RTL paragraph must keep its base direction on continuation
// lines: dropping it resolved bidi with an LTR base and put trailing
// punctuation on the wrong side of every wrapped line.
func TestOrderVisualBaseDirectionMatters(t *testing.T) {
	rtl := ""
	for _, p := range OrderVisual([]string{"שלום!"}, "rtl") {
		rtl += p.Text
	}
	none := ""
	for _, p := range OrderVisual([]string{"שלום!"}, "") {
		none += p.Text
	}
	if rtl == none {
		t.Skip("base direction does not change this input; pick a stronger probe")
	}
	// With an RTL base the bang leads in LTR draw order (visual left).
	if []rune(rtl)[0] != '!' {
		t.Fatalf("rtl base: got %q, want the bang at visual left", rtl)
	}
}

File: packages/text/src/__tests__/bidi.test.ts
// Bidirectional ordering (F38 FR-10). The format has carried
// `ParagraphStyle.direction` since the text model was written and nothing read
// it, so every right-to-left paragraph laid out left to right.
import { describe, expect, it } from "vitest";
import { resolveBaseDirection, resolveLevels, reorderRuns, orderLinePieces, hasRtl } from "../bidi";

const HE = "שלום"; // "shalom"
const AR = "مرحبا"; // "marhaba"

describe("resolveBaseDirection", () => {
  it("takes the first strong character when the paragraph says auto", () => {
    expect(resolveBaseDirection("hello world")).toBe("ltr");
    expect(resolveBaseDirection(HE)).toBe("rtl");
    expect(resolveBaseDirection(AR)).toBe("rtl");
    // Leading punctuation, digits and spaces are not strong.
    expect(resolveBaseDirection(`  "123" ${HE}`)).toBe("rtl");
    expect(resolveBaseDirection(`  "123" hello`)).toBe("ltr");
    // A Latin word before the Hebrew makes the paragraph left-to-right.
    expect(resolveBaseDirection(`Re: ${HE}`)).toBe("ltr");
  });

  it("honours an explicit declaration over the content", () => {
    expect(resolveBaseDirection(HE, "ltr")).toBe("ltr");
    expect(resolveBaseDirection("hello", "rtl")).toBe("rtl");
  });

  it("defaults to left-to-right with no strong character at all", () => {
    expect(resolveBaseDirection("")).toBe("ltr");
    expect(resolveBaseDirection("123 456 ...")).toBe("ltr");
  });
});

describe("resolveLevels", () => {
  it("gives right-to-left text an odd level inside a left-to-right paragraph", () => {
    const text = `hi ${HE}`;
    const levels = resolveLevels(text, "ltr");
    expect(levels).toHaveLength(text.length);
    expect(levels[0]).toBe(0); // "h"
    expect(levels[text.length - 1] % 2).toBe(1); // Hebrew
  });

  it("gives Latin an even level inside a right-to-left paragraph", () => {
    const text = `${AR} ok`;
    const levels = resolveLevels(text, "rtl");
    expect(levels[0] % 2).toBe(1); // Arabic at base level 1
    expect(levels[text.length - 1] % 2).toBe(0); // "k" raised to 2
  });

  it("keeps European digits running left-to-right inside Arabic (W2, I2)", () => {
    // Digits after Arabic become Arabic-Indic in class but still read LTR, so
    // they must land on an even level above the base.
    const text = `${AR} 2026`;
    const levels = resolveLevels(text, "rtl");
    const digitLevel = levels[text.length - 1];
    expect(digitLevel % 2).toBe(0);
    expect(digitLevel).toBeGreaterThan(1);
  });

  it("resolves a neutral between two right-to-left runs to right-to-left (N1)", () => {
    const text = `${HE} - ${HE}`;
    const levels = resolveLevels(text, "ltr");
    const dashIndex = text.indexOf("-");
    expect(levels[dashIndex] % 2).toBe(1);
  });

  it("resolves a neutral between opposite directions to the base (N2)", () => {
    const text = `abc - ${HE}`;
    const levels = resolveLevels(text, "ltr");
    expect(levels[text.indexOf("-")] % 2).toBe(0);
  });

  it("counts one level per code unit so callers can slice the string", () => {
    const emoji = "a😀b"; // the emoji is a surrogate pair
    expect(resolveLevels(emoji, "ltr")).toHaveLength(emoji.length);
  });
});

describe("reorderRuns", () => {
  it("reverses a right-to-left run inside left-to-right text", () => {
    // levels: L L R R L  ->  the middle pair swaps places as a block
    const runs = reorderRuns([0, 0, 1, 1, 0]);
    expect(runs.map((r) => [r.start, r.end, r.level])).toEqual([
      [0, 2, 0],
      [2, 4, 1],
      [4, 5, 0],
    ]);
  });

  it("reverses the whole line for a right-to-left paragraph", () => {
    // A right-to-left base with an embedded Latin word: the Latin keeps its
    // own left-to-right order but sits to the left of the Arabic.
    const runs = reorderRuns([1, 1, 2, 2, 1]);
    expect(runs.map((r) => r.level)).toEqual([1, 2, 1]);
    expect(runs[0].start).toBe(4); // the last logical run is displayed first
    expect(runs[2].start).toBe(0);
  });

  it("leaves a purely left-to-right line untouched", () => {
    expect(reorderRuns([0, 0, 0]).map((r) => [r.start, r.end])).toEqual([[0, 3]]);
  });
});

describe("orderLinePieces", () => {
  const seg = (text: string, id: string) => ({ text, id });

  it("leaves left-to-right runs in authored order", () => {
    const pieces = orderLinePieces([seg("Hello ", "a"), seg("world", "b")], "ltr");
    expect(pieces.map((p) => p.text)).toEqual(["Hello ", "world"]);
  });

  it("reverses the run order of a right-to-left paragraph", () => {
    // Two style runs, both Hebrew: displayed right to left, so the SECOND
    // authored run is drawn first (leftmost). This is the bug that made a
    // styled Hebrew sentence unreadable while a single unstyled run looked fine.
    const pieces = orderLinePieces([seg("שלום ", "a"), seg("עולם", "b")], "rtl");
    expect(pieces.map((p) => p.item.id)).toEqual(["b", "a"]);
  });

  it("keeps an embedded Latin phrase readable inside Arabic", () => {
    const pieces = orderLinePieces([seg(`${AR} `, "a"), seg("HyCanvas", "b")], "rtl");
    // The Latin run is drawn leftmost, unreversed, with the Arabic to its right.
    expect(pieces[0].item.id).toBe("b");
    expect(pieces[0].text).toBe("HyCanvas");
    expect(pieces[0].level % 2).toBe(0);
    expect(pieces[pieces.length - 1].item.id).toBe("a");
  });

  it("keeps Hebrew to the right of a Latin lead-in", () => {
    const pieces = orderLinePieces([seg("Note: ", "a"), seg(HE, "b")], "ltr");
    expect(pieces.map((p) => p.item.id)).toEqual(["a", "b"]);
  });

  it("splits one style run across two levels when it mixes directions", () => {
    const pieces = orderLinePieces([seg(`abc ${HE} def`, "only")], "ltr");
    // Three pieces: Latin, Hebrew at an odd level, Latin.
    expect(pieces).toHaveLength(3);
    expect(pieces.map((p) => p.level % 2)).toEqual([0, 1, 0]);
    // Every piece still points at the run it came from, so styling survives.
    expect(new Set(pieces.map((p) => p.item.id))).toEqual(new Set(["only"]));
    // A SINGLE embedded run reverses to itself, so the concatenation is
    // unchanged here; what changes is that the middle piece is marked
    // right-to-left for the renderer. Two embedded runs do reorder.
    expect(pieces.map((p) => p.text).join("")).toBe(`abc ${HE} def`);
  });

  it("reorders two right-to-left words around a left-to-right one", () => {
    // Logical: HE1 "and" HE2 in a right-to-left paragraph. Displayed right to
    // left, HE1 is rightmost, so the pieces come out reversed.
    const pieces = orderLinePieces(
      [seg(`${HE} `, "he1"), seg("and", "en"), seg(` ${AR}`, "ar")],
      "rtl",
    );
    expect(pieces.map((p) => p.item.id)).toEqual(["ar", "en", "he1"]);
  });

  it("preserves the full text content regardless of ordering", () => {
    const items = [seg("Total: ", "a"), seg("42 ", "b"), seg(AR, "c")];
    for (const base of ["ltr", "rtl"] as const) {
      const joined = orderLinePieces(items, base)
        .map((p) => p.text)
        .join("");
      const original = items.map((i) => i.text).join("");
      expect([...joined].sort().join("")).toBe([...original].sort().join(""));
    }
  });

  it("handles an empty line and empty runs", () => {
    expect(orderLinePieces([], "rtl")).toEqual([]);
    expect(orderLinePieces([seg("", "a")], "rtl").map((p) => p.text).join("")).toBe("");
  });
});

describe("hasRtl", () => {
  it("detects Hebrew and Arabic, and ignores digits and Latin", () => {
    expect(hasRtl("hello")).toBe(false);
    expect(hasRtl("12345 .,")).toBe(false);
    expect(hasRtl(HE)).toBe(true);
    expect(hasRtl(`mixed ${AR}`)).toBe(true);
  });
});

// The layout engine must actually consult the direction. This is the
// end-to-end version of the bug: the format carried `direction` and the
// engine ignored it.
import { layoutText } from "../layout";
import type { TextNode } from "@hc/schema";

function node(paragraphs: { text: string; direction?: "ltr" | "rtl" | "auto"; align?: "left" | "right" }[]): TextNode {
  return {
    id: "t",
    type: "text",
    transform: { x: 0, y: 0, scaleX: 1, scaleY: 1, rotation: 0 },
    size: { width: 400, height: 100 },
    box: { mode: "fixed", width: 400, height: 100, autoFit: { enabled: false, min: 6, max: 512 }, verticalAlign: "top" },
    content: paragraphs.map((p) => ({
      runs: [{ text: p.text, style: { fontFamily: "system", fontStyle: "Regular", fontSize: 16 } }],
      style: { align: p.align, direction: p.direction ?? "auto" },
    })),
  } as unknown as TextNode;
}

describe("layoutText direction", () => {
  it("resolves and reports the base direction per paragraph", () => {
    const res = layoutText(node([{ text: "hello" }, { text: HE }]));
    expect(res.lines[0].direction).toBe("ltr");
    expect(res.lines[res.lines.length - 1].direction).toBe("rtl");
  });

  it("right-aligns a right-to-left paragraph that never chose an alignment", () => {
    const res = layoutText(node([{ text: AR }]));
    expect(res.lines[0].align).toBe("right");
  });

  it("keeps an explicit alignment even when the text is right-to-left", () => {
    const res = layoutText(node([{ text: AR, align: "left" }]));
    expect(res.lines[0].align).toBe("left");
  });

  it("leaves left-to-right paragraphs exactly as before", () => {
    const res = layoutText(node([{ text: "hello world" }]));
    expect(res.lines[0].align).toBe("left");
    expect(res.lines[0].direction).toBe("ltr");
    expect(res.lines[0].segments.map((s) => s.text).join("")).toBe("hello world");
  });
});
